pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -0,0 +1,347 @@
+
+// SingleNUMAPolicy is implementation of the policy interface for the single NUMA policy
+// TODO: need to re-evaluate what field we really need
+type singleNUMAPolicy struct {
+	// machineInfo contains machine memory related information
+	machineInfo *cadvisorapi.MachineInfo
+	// reserved contains memory that reserved for kube
+	systemReserved ReservedMemory
+	// topology manager reference to get container Topology affinity
+	affinity topologymanager.Store
+}
+
+var _ Policy = &singleNUMAPolicy{}
+
+// NewPolicySingleNUMA returns new single NUMA policy instance
+func NewPolicySingleNUMA(machineInfo *cadvisorapi.MachineInfo, reserved ReservedMemory, affinity topologymanager.Store) Policy {
+	// TODO: check if we have enough reserved memory for the system
+	//for _, node := range reserved {
+	//	if node[v1.ResourceMemory] == 0 {
+	//
+	//	}
+	//}
+	return &singleNUMAPolicy{
+		machineInfo:    machineInfo,
+		systemReserved: reserved,
+		affinity:       affinity,
+	}
+}
+
+func (p *singleNUMAPolicy) Name() string {
+	return string(policyTypeSingleNUMA)
+}
+
+func (p *singleNUMAPolicy) Start(s state.State) error {
+	if err := p.validateState(s); err != nil {
+		klog.Errorf("[memorymanager] Invalid state: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	return nil
+}
+
+// Allocate call is idempotent
+func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) error {
+	// allocate the memory only for guaranteed pods
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	klog.Infof("[memorymanager] Allocate (pod: %s, container: %s)", pod.Name, container.Name)
+	if blocks := s.GetMemoryBlocks(string(pod.UID), container.Name); blocks != nil {
+		klog.Infof("[memorymanager] Container already present in state, skipping (pod: %s, container: %s)", pod.Name, container.Name)
+		return nil
+	}
+
+	// Call Topology Manager to get the aligned memory affinity across all hint providers.
+	hint := p.affinity.GetAffinity(string(pod.UID), container.Name)
+	klog.Infof("[memorymanager] Pod %v, Container %v Topology Affinity is: %v", pod.UID, container.Name, hint)
+
+	// TODO: should we pin the memory once the affinity hint is not available?
+	// we can pin the memory to a NUMA node different from one, that CPU manager assigned CPU's from
+	// so it can degrade container performance
+	if hint.NUMANodeAffinity == nil {
+		return fmt.Errorf("[memorymanager] failed to get NUMA affinity")
+	}
+
+	affinityBits := hint.NUMANodeAffinity.GetBits()
+	if len(affinityBits) > 1 {
+		return fmt.Errorf("[memorymanager] NUMA affinity has more than one NUMA node")
+	}
+
+	machineState := s.GetMachineState()
+
+	var containerBlocks []state.Block
+	for resourceName, q := range container.Resources.Requests {
+		if resourceName != v1.ResourceMemory && !corehelper.IsHugePageResourceName(resourceName) {
+			continue
+		}
+
+		nodeResourceState := machineState[affinityBits[0]][resourceName]
```

Comment:
I think we can merge affinityBits to single affinityBit.
Then return error if the single affinityBit indicates more than one NUMA node in somewhere neer line 104.

And we can use the single affinityBit on line 117 rather than taking first affinityBit from affinityBits.

Reply:
I already have this check
```
if len(affinityBits) > 1 {
		return fmt.Errorf("[memorymanager] NUMA affinity has more than one NUMA node")
```
Does it not enough?